arch/rv32i: allow full address range to be covered by `{TOR,NAPOT}RegionSpec`

lschuermann opened the pull request:
### Pull Request Overview

This changes the TORRegionSpec and NAPOTRegionSpec struct's internal representation to store pre-generated `pmpaddrX` CSR values, which removes artificial restrictions to them covering only addresses in `[0; usize::MAX)` (that is, excluding address 2**XLEN - 1) due to the prior internal representation using an (addr, len) specification.

It retains the constructors---albeit renamed to clarify their semantics---to generate such a pmpaddr value from a size and address. It also adds a constructor to build a TORRegionSpec and NAPOTRegionSpec directly from a (pair of) raw pmpaddrX CSR values. These constructors still uphold the same invariants over these types, but need to perform less checks due to the nature of the `pmpaddrX` CSR encoding.

It removes the `addr` and `len` methods that these types previously had, as they now can return values out of range for the `usize` integer type. Convenience-methods to introspect the `RegionSpec` structs apart from retrieving the `pmpaddrX` CSR value may be re-added later, either with using a 64-bit integer type (which can hold all legal addresses and lengths even for 64-bit platforms), or alternative representations. This is not currently required.

The change also adds an additional convenience method to create a `NAPOTRegionSpec` from a pair of start and end addresses, removing pointer arithmetic in the board's initialization routines.


Additionally, RV64 platforms support only a 56 bit physical address space. For this reason (and because addresses in `pmpaddrX` CSRs are left-shifted by 2 bit) the uppermost 10 bits of a `pmpaddrX` CSR are defined as WARL-0. To avoid unexpected behavior when configuring the PMP with such violating addresses, a second commit introduces checks that any `pmpaddrX` values represented and encoded by TORRegionSpec or NAPOTRegionSpec instances maintain the above invariant.


Fixes #4333.

### Testing Strategy

This pull request needs testing.


### TODO or Help Wanted

I still want to write unit tests for this code.


### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

lschuermann commented:
I will want to add unit tests to this code, but would first like to hear from @ZhekaS (and anyone else interested!) whether this approach seems reasonable.

ZhekaS commented:
> I will want to add unit tests to this code, but would first like to hear from @ZhekaS (and anyone else interested!) whether this approach seems reasonable.

I feel like having a convenience constructor `from_log2size(exp: u8)` would be more user-friendly. NAPOT address representation is quite confusing, so it's composition better be "automated".

lschuermann commented:
> I feel like having a convenience constructor from_log2size(exp: u8) would be more user-friendly. NAPOT address representation is quite confusing, so it's composition better be "automated".

I personally don't find this particularly intuitive, because you'd not just have the region size (expressed as a logarithm), but also a base address. IIRC the relevant constraints with NAPOT regions are that their size must be 8 bytes, and their start address mod their size == 0. Thus the constructor would be:

`from_start_log2size(start: *const (), size_exp: u8)`

...and that just seems harder to use than start + size to me.

On the other hand, we already have two constructors, so as long as it's tested I won't stand in the way of adding a third.

@ZhekaS Let me know what you think.

alistair23 commented on chips/qemu_rv32_virt_chip/src/uart.rs:
```diff
@@ -249,7 +249,7 @@ impl Uart16550<'_> {
         // Check if the register contained a valid interrupt at all
         if !iir.matches_all(IIR::Pending::Pending) {
-            panic!("UART 16550: interrupt without interrupt");
+            panic!("UART 16550: interrupt without interrupt: {:?}", iir);
```
I think this slipped in by accident

lschuermann replied:
Indeed!

alistair23 commented:
LGTM. I think storing the pmpaddr values directly is a good plane. Adding a `full_address_space()` also seems like a handy helper to have around.

I think @ZhekaS idea of having `end` being the last address of the region is error prone as it isn't going to be obvious to callers what `end` should be and I see a few off by one mistakes being made in the future.

jrvanwhy commented on arch/rv32i/src/pmp.rs:
```diff
@@ -105,53 +115,77 @@ impl From<mpu::Permissions> for TORUserPMPCFG {
-    pub fn size(&self) -> usize {
-        self.size
+        Self::from_pmpaddr_csr(
+            ((start as usize) + (size - 1).overflowing_shr(1).0)
```
Now that our MSRV is 1.84, we can replace `start as usize` with `start.addr()`, which is more correct and IMO is clearer.

jrvanwhy commented on arch/rv32i/src/pmp.rs:
```diff
@@ -105,53 +115,77 @@ impl From<mpu::Permissions> for TORUserPMPCFG {
-    /// Retrieve the end address of this [`NAPOTRegionSpec`].
-    pub fn end(&self) -> *const u8 {
-        unsafe { self.start.add(self.size) }
+    /// Construct a new [`NAPOTRegionSpec`] from a start .
```
This sentence isn't complete (I think it was intended to say "from a start and end address.".

jrvanwhy commented on arch/rv32i/src/pmp.rs:
```diff
@@ -105,53 +115,77 @@ impl From<mpu::Permissions> for TORUserPMPCFG {
+    /// [`NAPOTRegionSpec`]'s documentation are satisfied, otherwise `None`.
+    pub fn from_start_end(start: *const u8, end: *const u8) -> Option<Self> {
+        (end as usize)
+            .checked_sub(start as usize)
```
Same as before, as of Rust 1.84 I think using `.addr()` is clearer than `as usize`.

jrvanwhy commented on arch/rv32i/src/pmp.rs:
```diff
@@ -105,53 +115,77 @@ impl From<mpu::Permissions> for TORUserPMPCFG {
+    ///
+    /// This method accepts a `start` and `end` address. It returns
+    /// `Some(region)` when all constraints specified in the
+    /// [`NAPOTRegionSpec`]'s documentation are satisfied, otherwise `None`.
```
I'm guessing that `start` is inclusive and `end` is exclusive, like `a..b` ranges in Rust? This comment should document that.

jrvanwhy commented on arch/rv32i/src/pmp.rs:
```diff
@@ -164,41 +198,80 @@ impl NAPOTRegionSpec {
+            return None;
         }
+
+        Self::from_pmpaddr_csrs((start as usize) >> 2, (end as usize) >> 2)
```
`.addr()` here as well.

lschuermann commented:
I addressed the requested comments and added tests for some of the constructors of `NAPOTRegionSpec` and `TORRegionSpec`, which work on both 32-bit and 64-bit hosts.

This actually took me much more time than expected, and this PR has been lingering for much too long. So I vouch for either waiting for the remaining tests for the `NAPOTRegionSpec::from_start_end` and `from_start_size` constructors, or even merging this as is.

> Another option might be to (regardless of addressing type) to let it take start and end, where end means the last address of the region. This way we both preserve the naming corresponding to the meaning and avoid any off-by-one issues.
> Here how the address calculations might look like:

While I think that your concern is reasonable and more (well-tested) constructors could be a good quality of life improvement, I'd maybe kindly ask you to implement these on top of this PR. I want to see this pushed over the finish line and not delaying it further, so those additions are best suited for a follow-up PR. :sweat_smile:
